display_vk/plugin.cpp

```diff
@@ -0,0 +1,199 @@
+#define GLFW_INCLUDE_VULKAN
+#include <GLFW/glfw3.h>
+
+#define VMA_IMPLEMENTATION
+#include "common/data_format.hpp"
+#include "common/phonebook.hpp"
+#include "common/plugin.hpp"
+#include "common/switchboard.hpp"
+#include "common/vk_util/display_sink.hpp"
+#include "common/vk_util/vulkan_utils.hpp"
+
+using namespace ILLIXR;
+
+class display_vk : public display_sink {
+public:
+    display_vk(const phonebook* const pb)
+        : sb{pb->lookup_impl<switchboard>()} { }
+
+    /**
+     * @brief This function sets up the GLFW and Vulkan environments. See display_sink::setup().
+     */
+    void setup() {
+        setup_glfw();
+        setup_vk();
+    }
+
+    /**
+     * @brief This function recreates the Vulkan swapchain. See display_sink::recreate_swapchain().
+     */
+    virtual void recreate_swapchain() override {
+        vkb::SwapchainBuilder swapchain_builder{vkb_device};
+        auto                  swapchain_ret = swapchain_builder.set_old_swapchain(vk_swapchain)
+                                 .set_desired_present_mode(VK_PRESENT_MODE_FIFO_KHR)
+                                 .set_desired_extent(display_params::width_pixels, display_params::height_pixels)
+                                 .build();
+        if (!swapchain_ret) {
+            ILLIXR::abort("Failed to create Vulkan swapchain. Error: " + swapchain_ret.error().message());
+        }
+        vkb_swapchain          = swapchain_ret.value();
+        vk_swapchain           = vkb_swapchain.swapchain;
+        swapchain_images       = vkb_swapchain.get_images().value();
+        swapchain_image_views  = vkb_swapchain.get_image_views().value();
+        swapchain_image_format = vkb_swapchain.image_format;
+        swapchain_extent       = vkb_swapchain.extent;
+    }
+
+    /**
+     * @brief This function polls GLFW events. See display_sink::poll_window_events().
+     */
+    void poll_window_events() override {
+        glfwPollEvents();
```

Comment:
Not sure if others have had the same issue as me, but I'm getting an issue here with:
```
X Error of failed request:  BadMatch (invalid parameter attributes)
  Major opcode of failed request:  149 ()
  Minor opcode of failed request:  4
  Serial number of failed request:  230
  Current serial number in output stream:  238
```
I'm not _quite_ sure what the issue is, but I'm guessing it has to do with the fact that [glfwPolllEvents() can only be called from the main thread](https://www.glfw.org/docs/3.3/group__window.html#ga37bd57223967b4211d60ca1a0bf3c832]. Running `gdb`, it seems like the thread that calls `setup_glfw()` is different from the thread that ends up calling `poll_window_events()`. I think the easiest workaround here is to just remove this function call (i.e. this function is no-op) because it doesn't seem like we're doing anything with the glfw events anyways.

Reply:
On some DEs such as Gnome, the window would freeze and gray out if we don't poll events. I'll look into the error.